src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactoryHelpers.cs

```diff
@@ -0,0 +1,345 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Listeners;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Microsoft.Azure.WebJobs.Host.Triggers;
+using Microsoft.Azure.WebJobs.Host.Bindings.Path;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+    /// <summary>
+    /// Helpers for creating bindings to common patterns such as Messaging or Streams.
+    /// This will add additional adapters to connect the user's parameter type to an IAsyncCollector. 
+    /// It will also see <see cref="IConverterManager"/> to convert 
+    /// from the user's type to the underlying IAsyncCollector's message type.
+    /// For example, for messaging patterns, if the user is a ICollector, this will add an adapter that implements ICollector and calls IAsyncCollector.  
+    /// </summary>
+    internal static class BindingFactoryHelpers
+    {
+         // Bind a trigger argument to various parameter types. 
+        // Handles either T or T[], 
+        internal static ITriggerDataArgumentBinding<TTriggerValue> GetTriggerArgumentBinding<TMessage, TTriggerValue>(
+            ITriggerBindingStrategy<TMessage, TTriggerValue> bindingStrategy,
+            ParameterInfo parameter,
+            IConverterManager converterManager,
+            out bool singleDispatch)
+        {
+            ITriggerDataArgumentBinding<TTriggerValue> argumentBinding = null;
+            if (parameter.ParameterType.IsArray)
+            {
+                // dispatch the entire batch in a single call. 
+                singleDispatch = false;
+
+                var elementType = parameter.ParameterType.GetElementType();
+
+                var innerArgumentBinding = GetTriggerArgumentElementBinding<TMessage, TTriggerValue>(elementType, bindingStrategy, converterManager);
+
+                argumentBinding = new ArrayTriggerArgumentBinding<TMessage, TTriggerValue>(bindingStrategy, innerArgumentBinding, converterManager);
+
+                return argumentBinding;
+            }
+            else
+            {
+                // Dispatch each item one at a time
+                singleDispatch = true;
+
+                var elementType = parameter.ParameterType;
+                argumentBinding = GetTriggerArgumentElementBinding<TMessage, TTriggerValue>(elementType, bindingStrategy, converterManager);
+                return argumentBinding;
+            }
+        }
+
+        // Bind a T. 
+        private static SimpleTriggerArgumentBinding<TMessage, TTriggerValue> GetTriggerArgumentElementBinding<TMessage, TTriggerValue>(
```

Comment:
The triggering stuff here is specifically for a message-pattern. TimerTrigger wouldn't use it. 

---
In reply to: [60989783](https://github.com/Azure/azure-webjobs-sdk/pull/694#discussion_r60989783) [](ancestors = 60989783)